### ISSUE TITLE
RAM-efficient shuffling in dataset

### ISSUE THREAD

**Issue by windo** [tensorflow/tensorflow#24351]
**System information**
- TensorFlow version (you are using): 1.12.0
- Are you willing to contribute it (Yes/No): Maybe, if I get some direction

**Describe the feature and the current behavior/state.**

Currently to shuffle datasets, `shuffle()` needs to read `shuffle_buffer` worth of records into memory. If the records are ~large and the dataset contains ordered/consequitve/correlated records then the user may need to pre-shuffle and/or split the data and/or commmit a good chunk of ram for shuffling the records sufficiently to avoid overfitting to features present in the particular part of the dataset.

It would be possible to trade off reading the dataset just once in order to shuffle more efficiently in terms of memory usage. For example, the ram-efficient dataset shuffler could read the dataset multiple times starting from offsets `0 .. N` and skip over `N` records on each pass. 

**Will this change the current api? How?**

It would add a new method to `tf.data.Dataset`, or maybe an option/experiment for `shuffle()`.

**Who will benefit with this feature?**

Anyne training on sequential (I guess correleated) datasets where individual records are on large enough to become a limiting factor for large `shuffle_buffer` sizes.

**Comment by jsimsa** [comment]
IIUC, your proposal relies on the existence of an efficient implementation of skip / seek for the dataset to be shuffled. Unfortunately, not all dataset types meet this requirement. For instance, `TFRecordDataset` -- commonly used for storing data at rest -- does not support efficient skip/ seek and to seek to an offset, you need to sequentially read all the data up to that offset.

The good news is that you could achieve RAM-efficient shuffling using existing tf.data API:

```
# `dataset` stores the dataset to shuffle
dataset = dataset.window(N).flat_map(lambda x: x.shuffle(M))
```

This will partition the dataset into windows of size `N` and shuffle each of them using buffer of size `M`.
